imap/jmap_core.c

```diff
@@ -399,6 +415,171 @@ static int getblob_cb(const conv_guidrec_t* rec, void* vrock)
     return 0;
 }
 
+static const jmap_property_t blob_xprops[] = {
+    {
+        "data:asText",
+        NULL,
+        JMAP_PROP_SERVER_SET | JMAP_PROP_IMMUTABLE | JMAP_PROP_SKIP_GET
+    },
+    {
+        "data:asHex",
+        NULL,
+        JMAP_PROP_SERVER_SET | JMAP_PROP_IMMUTABLE | JMAP_PROP_SKIP_GET
+    },
+    {
+        "data:asBase64",
+        NULL,
+        JMAP_PROP_SERVER_SET | JMAP_PROP_IMMUTABLE | JMAP_PROP_SKIP_GET
+    },
+    {
+        "data",
+        NULL,
+        JMAP_PROP_SERVER_SET | JMAP_PROP_IMMUTABLE
+    },
+    {
+        "size",
+        NULL,
+        JMAP_PROP_SERVER_SET | JMAP_PROP_IMMUTABLE
+    },
+    { NULL, NULL, 0 }
+};
+
+struct blob_xrange {
+    size_t offset;
+    size_t length;
+};
+
+static int _parse_xrange(jmap_req_t *req __attribute__((unused)),
+                         struct jmap_parser *parser,
+                         const char *key,
+                         json_t *arg,
+                         void *rock)
+{
+    struct blob_xrange *rangep = rock;
+
+    if (!strcmp(key, "offset")) {
+        if (json_is_integer(arg))
+            rangep->offset = json_integer_value(arg);
+        else
+            jmap_parser_invalid(parser, "offset");
+        return 1;
+    }
+
+    if (!strcmp(key, "length")) {
+        if (json_is_integer(arg))
+            rangep->length = json_integer_value(arg);
+        else
+            jmap_parser_invalid(parser, "length");
+        return 1;
+    }
+
+    return 0;
+}
+
+static int jmap_blob_xget(jmap_req_t *req)
+{
+    struct jmap_parser parser = JMAP_PARSER_INITIALIZER;
+    struct jmap_get get;
+    json_t *err = NULL;
+    json_t *jval;
+    size_t i;
+
+    /* Parse request */
+    struct blob_xrange range = { 0, 0 };
+    jmap_get_parse(req, &parser, blob_xprops, /*allow_null_ids*/0,
+                   &_parse_xrange, &range, &get, &err);
+    if (err) {
+        jmap_error(req, err);
+        goto done;
+    }
+
+    /* Sort blob lookups by mailbox */
+    json_array_foreach(get.ids, i, jval) {
+        const char *blob_id = json_string_value(jval);
+        if (*blob_id == 'G') {
+            jmap_getblob_context_t ctx;
+            jmap_getblob_ctx_init(&ctx, req->accountid, blob_id, NULL, 1);
+            int r = jmap_getblob(req, &ctx);
+            if (r) {
+                json_array_append_new(get.not_found, json_string(blob_id));
+            }
+            else {
+                const char *base = buf_base(&ctx.blob);
+                size_t len = buf_len(&ctx.blob);
+                if (range.offset <= len) {
+                    base += range.offset;
+                    len -= range.offset;
+                }
+                else {
+                    r = -1;
```

Comment:
I assume that's copied from old code, still I think that `-1` isn't a good choice here. Instead it should be set to HTTP_SERVER_ERROR or any more appropriate error. In this particular case wouldn't it even result in an `invalidArguments` JMAP error?

Remark: for new JMAP methods, I even tend to make them return `void` and take a `json_t **err` argument. That way I force the method handler to take care or error reporting.

Reply:
My plan here was that if there's an error, we add the blob to notFound, on the grounds that there was no blob data found within the specified range.  It would also be possible to just return `null` for the data bits if the range doesn't intersect but the blob does exist - I think that's reasonable.